docs(app): clarify iOS setup instructions, especially around use_frameworks (#6856)
invertase/react-native-firebase: docs/index.md

To do this, open your `/ios/{projectName}/AppDelegate.mm` file (or `AppDelegate.m` if on older react-native), and add the following:

> At the top of the file, import the Firebase SDK right after `'#import "AppDelegate.h"'`:

Reviewer: This will likely help people who mistakenly put it inside a compiler IFDEF conditional, that happens frequently. Good suggestion

```
- (BOOL)application:(UIApplication *)application didFinishLaunchingWithOptions:(NSDictionary *)launchOptions {
  // Add me --- \/

>   if ([FIRApp defaultApp] == nil) { [FIRApp configure]; }

Reviewer: This "check for null / only configure if defaultApp is null" has been the source of some debate, I'm not sure it is strictly required and I should probably remove it from my demo script.

Can you test if it works just fine as currently documented, that is just a straight `[FIRApp configure];` call?

Author: @mikehardy 
it works with `[FIRApp configure];`
so I will change this as currently documented!

Beginning with firebase-ios-sdk v9+ (react-native-firebase v15+) you must tell CocoaPods to use frameworks.

> Open the file `./ios/Podfile` and add this line inside your targets (right after `'config = use_native_modules!'`):

Reviewer: This is unstable advice unfortunately, as it changes from react-native version to react-native version. Perhaps something a little less specific code and a little more general 🤔 

```suggestion
Open the file `./ios/Podfile` and add this line inside your targets (right after the line calling the react native Podfile function to get the native modules config):
```

```ruby

> use_frameworks! :linkage => :static

Reviewer: Really good suggestion, linkage does need to be static for hermes to work

```

To use Static Frameworks on iOS, you also need to manually enable this for the project with the following global to your `/ios/Podfile` file:

```ruby
# right after `use_frameworks! :linkage => :static`
$RNFirebaseAsStaticFramework = true
```

> For build setting, add below codes inside your targets (right after `'__apply_Xcode_12_5_M1_post_install_workaround(installer)'`)

Reviewer: This part is dangerous - that workaround will likely go away soon. It is already empty inside the react-native scripts, and scheduled for removal I think

```ruby

>   installer.pods_project.targets.each do |target|
>     target.build_configurations.each do |config|
>       config.build_settings["GCC_WARN_INHIBIT_ALL_WARNINGS"] = "YES"
>     end
>   end

Reviewer: This is a matter of taste and must not be suggested as a general configuration for anyone
In general warnings are useful, they just aren't in my build demo context

```suggestion
```

>   installer.pods_project.targets.each do |target|
>     target.build_configurations.each do |config|
>       config.build_settings["CC"] = "clang"
>       config.build_settings["LD"] = "clang"
>       config.build_settings["CXX"] = "clang++"
>       config.build_settings["LDPLUSPLUS"] = "clang++"
>     end
>   end

Reviewer: This is only required for one specific method of doing the performance optimization I worked on here: https://reactnative.dev/docs/build-speed#xcode-specific-setup

It cannot be recommended as a general configuration for people because a) it is only a performance optimization and those are matters of opinion, while b) it opens a security risk because the compiler is now referenced via relative path, leading to potential for impersonating programs to be executed unexpectedly. I personally use it and I need it for the build demo script so I can run builds quickly but it should not be in react-native-firebase docs

```suggestion
```

  
  installer.aggregate_targets.each do |aggregate_target|
    aggregate_target.user_project.native_targets.each do |target|
      target.build_configurations.each do |config|
        config.build_settings['ONLY_ACTIVE_ARCH'] = 'YES'
        config.build_settings['EXCLUDED_ARCHS'] = 'i386'
      end
    end
    aggregate_target.user_project.save

>   end

Reviewer: I'm not sure this can be a general suggestion. I think this should be in the tips-and-tricks section instead https://rnfirebase.io/faqs-and-tips

There may be cases where a developer wants to build a non-active arch, and this eliminates that capability.

Author: After testing, it turned out that this was not an important part... I will delete it..

  installer.pods_project.targets.each do |target|
    if (target.name.eql?('FBReactNativeSpec'))
      target.build_phases.each do |build_phase|
        if (build_phase.respond_to?(:name) && build_phase.name.eql?('[CP-User] Generate Specs'))
          target.build_phases.move(build_phase, 0)
        end
      end
    end

>   end

Reviewer: Have you tried without this? I don't think it is required anymore - I have not had a chance to check yet:

https://github.com/facebook/react-native/issues/31149#issuecomment-1282465756

Author: you are right.. haha.. I will also delete it!

  
  installer.pods_project.targets.each do |target|
    target.build_configurations.each do |config|
      config.build_settings["ENABLE_BITCODE"] = "NO"
    end

>   end

Reviewer: This is a general issue, not a react-native-firebase issue. I don't believe this should be in our documentation - it should be (and is?I think?) documented elsewhere.

It could be a good tips-and-tricks addition though

Author: It works normally without modifying this part. I will delete the related content because it seems unnecessary.

```ruby
# Override Firebase SDK Version

> # Between `use_react_native!()` paragraph and `target 'RNFBTests' do ... end`

Reviewer: It does not matter where it goes actually. It's just defining a variable in Ruby syntax, at the top level.

```suggestion
```

> $FirebaseSDKVersion = '10.4.0'

Reviewer: As mentioned in another comment we are not compatible with 10.4.0 yet. This is an advanced configuration method anyway and basically no one should use it, definitely no one should use it long-term, as react-native-firebase should control the dependency version.

```suggestion
$FirebaseSDKVersion = '10.3.0'
```

Author: yeah I checked it also works with 10.3.0 too
thanks 👍
